src/backend/distributed/planner/merge_planner.c

```diff
@@ -701,13 +702,6 @@ ErrorIfRepartitionMergeNotSupported(Oid targetRelationId, Query *mergeQuery,
 							   "Postgres table is not yet supported")));
 	}
 
-	queryRteListProperties = GetRTEListPropertiesForQuery(sourceQuery);
```

Comment:
why are we removing the local table checks in this PR?

Reply:
With out this relaxation, we can't execute queries where '`source`' doesn't point to a direct reference to a table.

```
CREATE OR REPLACE FUNCTION src_tab_result()
RETURNS TABLE (id INT, val INT, bs bigint, tm timestamp) AS
$$
BEGIN
    RETURN QUERY SELECT src_tab.*, nextval('dest_tab_ser_seq'::regclass), now()::timestamp FROM src_tab;
END;
$$
LANGUAGE plpgsql;

MERGE INTO dest_tab t
USING (SELECT * FROM src_tab_result()) s
ON (s.id = t.tid)
    WHEN MATCHED THEN
        UPDATE SET val = t.val + 1
    WHEN NOT MATCHED THEN
        INSERT VALUES(s.id, s.val, s.bs, s.tm);

```

Reply:
By the way, in the repartition strategy, the plan must have a CitusExecNode at the top of the tree, hence the restriction. Now that we support the pull-to-coordinator strategy, strictly speaking, we don't need that restriction. Do you foresee any issues?

Reply:
>    WHEN NOT MATCHED THEN
	INSERT VALUES(s.sid, s.val, now());

Hmm, actually, we still cannot have functions in the MergeAction fields. We can only allow in the source.

Otherwise, if replication factor > 1, then we'd insert two different "now" values. We should fix that anyway

Reply:
That's right, we can't have volatile functions in MERGE actions

```
MERGE INTO target USING source ON (source.sid = target.tid)
        WHEN NOT MATCHED THEN
INSERT VALUES (random(), source.val);
```
You might end up inserting a wrong dist-key value in the target.

Reply:
>>because we may want to backport this to 12.0.X.

We have several improvements/enhancements across the versions. I would consider this as an improvement and not a backport. This, and maybe some more stuff, will end up in version 12.1, which is not far away. If there is a compelling reason, such as a significant customer request, we can consider it.

Reply:
>>I think that is still a limitation (e.g., functions / local tables etc)

Local tables-> You meant Citus local or Postgres tables?

Reply:
-- I guess, this is still not supported
-- probably get ERROR:  STABLE functions used in MERGE queries cannot be called with column references
-- but should be OK

That's what I get using your testcase, are you getting a different error?


```
MERGE INTO dest_tab t
USING (SELECT *, now() as nw FROM src_tab) s
ON (s.sid = t.tid)
    WHEN MATCHED THEN
        UPDATE SET val = t.val + 1
    WHEN NOT MATCHED THEN
        INSERT VALUES(s.sid, s.val, nw);
psql::22: ERROR:  STABLE functions used in MERGE queries cannot be called with column references

```

Reply:
@onderkalaci  I suspect you are using an incorrect branch?

`ERROR:  non-IMMUTABLE functions are not yet supported in MERGE sql with distributed tables`

That error is removed in this PR, could you please double check

Reply:
> Local tables-> You meant Citus local or Postgres tables?

Both. If we are going with this patch, we have to add bunch of regression tests where the source is [citus] local table(s) and the target distributed. 

I still don't think that is is a good idea to remove this check in this PR for (a) there is currently no user requirement for such a change (b) it increases the surface area of the patch quite a bit (c) it changes the focus of the patch, look we do not even have a test for the cases that we primarily try solving?

> . Do you foresee any issues?

No, should be fine overall. Mostly requires testing.

> @onderkalaci I suspect you are using an incorrect branch?

That error still exists in `MergeQualAndTargetListFunctionsSupported`? 

But I might have changed your branch while testing/playing with it, so you might see a different error. I think that was my expectation for that case.

> That's right, we can't have volatile functions in MERGE actions

I guess you have not worked on the fix yet -- or am I missing it?

Reply:
>>ERROR:  non-IMMUTABLE functions are not yet supported in MERGE sql with distributed tables

All your tests showed this error, which is not there in the merge-functions branch of this PR. Let's have quick sync-call on this, I've sent an invite.